File: Irrgarten/src/irrgarten/Labyrinth.java
package irrgarten;

import java.util.ArrayList;

/**
 * Represents a labyrinth where players and monsters can move around.
 */
public class Labyrinth {
    // Constants for characters used in the labyrinth
    public static final char BLOCK_CHAR = 'X';
    public static final char EMPTY_CHAR = '-';
    public static final char MONSTER_CHAR = 'M';
    public static final char COMBAT_CHAR= 'C';
    public static final char EXIT_CHAR = 'E';
    public static final int ROW = 0;
    public static final int COL = 1;

    // Labyrinth properties
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;

    // Matrices to hold monsters, players, and labyrinth layout
    private Monster[][] monsters;
    private Player[][] players;
    private char labyrinth[][];

    /**
     * Constructor to initialize the labyrinth with given dimensions and exit position.
     * @param nRows Number of rows in the labyrinth.
     * @param nCols Number of columns in the labyrinth.
     * @param exitRow Row index of the exit position.
     * @param exitCol Column index of the exit position.
     */
    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        this.nRows = nRows;
        this.nCols = nCols;
        
        // Initialize matrices
        this.monsters = new Monster[nRows][nCols];
        this.players = new Player[nRows][nCols];
        this.labyrinth =  new char[nRows][nCols];
        
        // Fill labyrinth with empty characters
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++){
                this.monsters[i][j] = null;
                this.players[i][j] = null;
                this.labyrinth[i][j] = Labyrinth.EMPTY_CHAR;
            }
        }

        // Set the exit position
        this.labyrinth[exitRow][exitCol] = Labyrinth.EXIT_CHAR;
    }

    /**
     * Spread players across the labyrinth.
     * @param players List of players to spread.
     */
    public void spreadPlayers(ArrayList<Player> players){
        for(Player p : players){
            int [] pos = this.randomEmptyPos();
            int nul_val = -1;
            putPlayer2D(nul_val,nul_val, pos[Labyrinth.ROW], pos[Labyrinth.COL], p);
        }
    }

    /**
     * Check if a player has reached the exit position.
     * @return True if a player has reached the exit, false otherwise.
     */
    public boolean haveAWinner(){
        return (this.players[this.exitRow][this.exitCol] != null);
    }

    /**
     * Convert the labyrinth to a string representation.
     * @return String representation of the labyrinth.
     */
    @Override
    public String toString(){
        String labyrinthString = "";
        for (int i = 0; i < this.nRows; i ++){
            labyrinthString += "|";
            for (int j = 0; j < this.nCols; j++){
                labyrinthString += this.labyrinth[i][j] + "|";
            }
            labyrinthString += "\n";
        }
        
        return labyrinthString;
    }


    /**
     * Adds a monster to the specified position in the labyrinth.
     * @param row The row index where the monster will be added.
     * @param col The column index where the monster will be added.
     * @param monster The monster to be added.
     */
    public void addMonster(int row, int col, Monster monster){
        if (this.posOK(row, col) && this.emptyPos(row,col)) {
            // Save a reference to the monster in the corresponding attributes
            this.monsters[row][col] = monster;
            // Set the monster character in the labyrinth
            this.labyrinth[row][col] = Labyrinth.MONSTER_CHAR;
            // Set the position of the monster
            monster.setPos(row, col);
        }
    }

    /**
     * Places a player in the labyrinth based on the given direction.
     * @param direction The direction in which the player should be placed.
     * @param player The player to be placed.
     * @return The monster encountered, if any, while moving the player.
     */
    public Monster putPlayer(Directions direction, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        int [] newPos = dir2Pos(oldRow, oldCol, direction);
        Monster monster = putPlayer2D(oldRow, oldCol, 
                                      newPos[Labyrinth.ROW], 
                                      newPos[Labyrinth.COL],player);
        return monster;
        
    }

    /**
     * Adds a block to the labyrinth at the specified position and orientation.
     * @param orientation The orientation of the block.
     * @param startRow The starting row index of the block.
     * @param startCol The starting column index of the block.
     * @param length The length of the block.
     */
    public void addBlock(Orientation orientation, int startRow, int startCol, int length) {
        int incRow, incCol;
        if(orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }
        else{
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow, col = startCol;
        
        while (this.posOK(row,col) && this.emptyPos(row,col) && (length > 0)){
            this.labyrinth[row][col] = Labyrinth.BLOCK_CHAR;
            length-=1;
            row+=incRow;
            col+=incCol;
        }
            
    }

    /**
     * Determines valid moves from the given position.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return A list of valid moves from the position.
     */
    public ArrayList<Directions> validMoves(int row, int col) {
        ArrayList<Directions> output = new ArrayList<Directions>(0);
        
        if (this.canStepOn(row+1,col))
            output.add(Directions.DOWN);
        
        if (this.canStepOn(row-1,col))
            output.add(Directions.UP);
        
        if (this.canStepOn(row,col+1))
            output.add(Directions.RIGHT);
        
        if (this.canStepOn(row,col-1))
            output.add(Directions.LEFT);
        
        return output;
    }

    /**
     * Checks if a position is within the bounds of the labyrinth.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return True if the position is within the labyrinth bounds, false otherwise.
     */
    private boolean posOK(int row, int col){
        return (0 <= row) && (row < this.nRows) && (0 <= col) && (col < this.nCols);
    }

    /**
     * Checks if a position in the labyrinth is empty.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return True if the position is empty, false otherwise.
     */
    private boolean emptyPos(int row, int col){
        return (this.labyrinth[row][col] == Labyrinth.EMPTY_CHAR);
    }

    /**
     * Checks if there is a monster at the specified position in the labyrinth.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return True if there is a monster at the position, false otherwise.
     */
    private boolean monsterPos(int row, int col){
	    return (this.labyrinth[row][col] == Labyrinth.MONSTER_CHAR);
    }

    /**
     * Checks if there is an exit at the specified position in the labyrinth.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return True if there is an exit at the position, false otherwise.
     */
    private boolean exitPos(int row, int col){
	    return (this.labyrinth[row][col] == Labyrinth.EXIT_CHAR);
    }

    /**
     * Checks if there is a combat position at the specified position in the labyrinth.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return True if there is a combat position at the position, false otherwise.
     */
    private boolean combatPos(int row, int col){
        return (this.labyrinth[row][col] == Labyrinth.COMBAT_CHAR);
    }

    /**
     * Checks if it is possible to step on the specified position in the labyrinth.
     * @param row The row index of the position.
     * @param col The column index of the position.
     * @return True if it is possible to step on the position, false otherwise.
     */
    private boolean canStepOn(int row, int col){
        boolean canStep = false;
        if(this.posOK(row, col)) {
            boolean empty = this.emptyPos(row, col);
            boolean monsterPos = this.monsterPos(row, col);
            boolean exit = this.exitPos(row, col);
            canStep = empty || monsterPos || exit;
        }
        return canStep;
    }

    /**
     * Updates the old position in the labyrinth after moving.
     * @param row The row index of the old position.
     * @param col The column index of the old position.
     */
    private void updateOldPos(int row, int col){
        if (this.posOK(row, col)) {
            if (this.combatPos(row, col))
                this.labyrinth[row][col] = Labyrinth.MONSTER_CHAR;
            else 
                this.labyrinth[row][col] = Labyrinth.EMPTY_CHAR;
            
        }
    }

    /**
     * Converts a position and direction to a new position in the labyrinth.
     * @param row The current row index.
     * @param col The current column index.
     * @param direction The direction of movement.
     * @return An array containing the new row and column indices.
     */
    private int[] dir2Pos(int row, int col, Directions direction) {
        int [] pos = {row,col};
        switch (direction){
            case LEFT: 
                pos[Labyrinth.COL]--;  break;
            case RIGHT:
                pos[Labyrinth.COL]++;  break;
            case UP:
                pos[Labyrinth.ROW]--;  break;
            case DOWN:
                pos[Labyrinth.ROW]++;  break; 
        }
        return pos;
    }
    
    /**
     * Finds a random empty position in the labyrinth.
     * @return An array containing the row and column indices of the random empty position.
     */
    private int[] randomEmptyPos(){
        int[] rPos = new int[2];
        int rRow = Dice.randomPos(this.nRows);
        int rCol = Dice.randomPos(this.nCols);

        while (!this.posOK(rRow, rCol) || !this.emptyPos(rRow, rCol)) {
            rRow = Dice.randomPos(this.nRows);
            rCol = Dice.randomPos(this.nCols);
        }

        rPos[ROW] = rRow;
        rPos[COL] = rCol;
        return rPos;
    }

    /**
     * Moves a player to the specified position in the labyrinth.
     * @param oldRow The row index of the player's old position.
     * @param oldCol The column index of the player's old position.
     * @param row The row index of the player's new position.
     * @param col The column index of the player's new position.
     * @param player The player to be moved.
     * @return The monster encountered, if any, while moving the player.
     */
    private Monster putPlayer2D (int oldRow, int oldCol, int row, int col, Player player){
        Monster output = null;
        if (this.canStepOn(row,col)){
            
            if(this.posOK(oldRow,oldCol)){
                Player p = this.players[oldRow][oldCol];
                if(p == player){
                    this.updateOldPos(oldRow,oldCol);
                    this.players[oldRow][oldCol] = null;
                }
            }
            if(this.monsterPos(row,col)){
                this.labyrinth[row][col]=COMBAT_CHAR;
                output = this.monsters[row][col];
            }
            else{
                char number = player.getNumber();
                this.labyrinth[row][col]=number;
            }
            
            this.players[row][col]=player;
            player.setPos(row,col);
        }
        
        return output;
    }
    
    /**
     * Updates the player in the labyrinth after moving.
     * @param player The player to be updated
     */
    public void updatePlayer(Player player){
        if(this.players[player.getRow()][player.getCol()]!= null)
           this.players[player.getRow()][player.getCol()] = player;
    }
}

File: Irrgarten_ruby/labyrinth.rb
#encoding:utf-8
require_relative 'dice'
require_relative 'orientation'
require_relative 'directions'
module Irrgarten
    # The Labyrinth class represents a maze game.
    class Labyrinth
        # The character used to represent a block in the labyrinth.
        @@BLOCK_CHAR = 'X'

        # The character used to represent an empty space in the labyrinth.
        @@EMPTY_CHAR = '-'

        # The character used to represent a monster in the labyrinth.
        @@MONSTER_CHAR = 'M'

        # The character used to represent a combat position in the labyrinth.
        @@COMBAT_CHAR = 'C'

        # The character used to represent the exit in the labyrinth.
        @@EXIT_CHAR = 'E'

        # The row index in a position array.
        @@ROW = 0

        # The column index in a position array.
        @@COL = 1

        # Initializes a new Labyrinth instance.
        #
        # @param n_rows [Integer] the number of rows in the labyrinth
        # @param n_cols [Integer] the number of columns in the labyrinth
        # @param exit_col [Integer] the column index of the exit
        # @param exit_row [Integer] the row index of the exit
        def initialize(n_rows, n_cols, exit_col, exit_row)
            @exit_row = exit_row
            @exit_col = exit_col
            @n_rows = n_rows
            @n_cols = n_cols
            @monsters = Array.new(@n_rows) { Array.new(@n_cols, nil) }
            @players = Array.new(@n_rows) { Array.new(@n_cols, nil) }
            @labyrinth = Array.new(@n_rows) { Array.new(@n_cols, @@EMPTY_CHAR) }
            @labyrinth[@exit_row][@exit_col] = @@EXIT_CHAR
        end

        # Spreads the players in the labyrinth.
        #
        # @param players [Array<Player>] the players to spread
        def spread_players(players)
            players.each do |p|
                pos = self.random_empty_pos
                self.put_player_2D(-1, -1, pos[@@ROW], pos[@@COL], p)
            end
            nil
        end

        # Checks if there is a winner in the labyrinth.
        #
        # @return [Boolean] true if there is a winner, false otherwise
        def have_a_winner
            @players[@exit_row][@exit_col] != nil
        end

        # Returns a string representation of the labyrinth.
        #
        # @return [String] the labyrinth as a string
        def to_s
            string = ''
            @labyrinth.each do |row|
                string << '|'
                row.each do |box|
                    string << "#{box}|"
                end
                string << "\n"
            end
            string
        end

        # Adds a monster to the labyrinth at the specified position.
        #
        # @param row [Integer] the row index of the monster's position
        # @param col [Integer] the column index of the monster's position
        # @param monster [Monster] the monster to add
        def add_monster(row, col, monster)
            if self.pos_OK(row, col) && self.empty_pos(row, col)
                @monsters[row][col] = monster
                @labyrinth[row][col] = @@MONSTER_CHAR
                monster.set_pos(row, col)
            end
        end

        # Puts a player in the labyrinth at the specified direction.
        #
        # @param direction [Symbol] the direction to put the player
        # @param player [Player] the player to put
        def put_player(direction, player)
            old_row = player.row
            old_col = player.col
            new_pos = self.dir2pos(old_row, old_col, direction)
            self.put_player_2D(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
        end

        # Adds a block to the labyrinth.
        #
        # @param orientation [Symbol] the orientation of the block (:horizontal or :vertical)
        # @param start_row [Integer] the row index of the block's starting position
        # @param start_col [Integer] the column index of the block's starting position
        # @param length [Integer] the length of the block
        def add_block(orientation, start_row, start_col, length)
            if orientation == Orientation::VERTICAL
                inc_row = 1
                inc_col = 0
            else
                inc_row = 0
                inc_col = 1
            end

            row = start_row
            col = start_col
            while self.pos_OK(row, col) && self.empty_pos(row, col) && length > 0
                @labyrinth[row][col] = @@BLOCK_CHAR
                length -= 1
                row += inc_row
                col += inc_col
            end
        end

        # Returns the valid moves from the specified position.
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Array<Directions>] the valid moves from the position
        def valid_moves(row, col)
            output = Array.new(0)
            output << Directions::DOWN if can_step_on(row + 1, col)
            output << Directions::UP if can_step_on(row - 1, col)
            output << Directions::RIGHT if can_step_on(row, col + 1)
            output << Directions::LEFT if can_step_on(row, col - 1)
            output
        end

        private

        # Checks if the specified position is within the labyrinth bounds.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Boolean] true if the position is within bounds, false otherwise
        def pos_OK(row, col)
            (0 <= row) && (row < @n_rows) && (0 <= col) && (col < @n_cols)
        end

        # Checks if the specified position is empty.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Boolean] true if the position is empty, false otherwise
        def empty_pos(row, col)
            @labyrinth[row][col] == @@EMPTY_CHAR
        end

        # Checks if the specified position contains a monster.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Boolean] true if the position contains a monster, false otherwise
        def monster_pos(row, col)
            @labyrinth[row][col] == @@MONSTER_CHAR
        end

        # Checks if the specified position is the exit.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Boolean] true if the position is the exit, false otherwise
        def exit_pos(row, col)
            @labyrinth[row][col] == @@EXIT_CHAR
        end

        # Checks if the specified position is a combat position.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Boolean] true if the position is a combat position, false otherwise
        def combat_pos(row, col)
            @labyrinth[row][col] == @@COMBAT_CHAR
        end

        # Checks if the specified position can be stepped on.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        # @return [Boolean] true if the position can be stepped on, false otherwise
        def can_step_on(row, col)
            valid_pos = self.pos_OK(row,col)
            valid_pos && (self.empty_pos(row, col) || self.monster_pos(row, col) || self.exit_pos(row, col))
        end

        # Updates the old position after moving.
        #
        # @param row [Integer] the row index of the position
        # @param col [Integer] the column index of the position
        def update_old_pos(row, col)
            if self.pos_OK(row, col)
                if self.combat_pos(row, col)
                    @labyrinth[row][col] = @@MONSTER_CHAR
                else
                    @labyrinth[row][col] = @@EMPTY_CHAR
                end
            end
        end

        # Converts a direction to a new position.
        #
        # @param row [Integer] the row index of the current position
        # @param col [Integer] the column index of the current position
        # @param direction [Symbol] the direction to move
        # @return [Array<Integer>] the new position as [row, col]
        def dir2pos(row, col, direction)
            pos = [row, col]
            case direction
            when Irrgarten::Directions::LEFT
                pos[@@COL] -= 1
            when Irrgarten::Directions::RIGHT
                pos[@@COL] += 1
            when Irrgarten::Directions::UP
                pos[@@ROW] -= 1
            when Irrgarten::Directions::DOWN
                pos[@@ROW] += 1
            end
            pos
        end

        # Returns a random empty position in the labyrinth.
        #
        # @return [Array<Integer>] the random empty position as [row, col]
        def random_empty_pos
            r_row = Dice.random_pos(@n_rows)
            r_col = Dice.random_pos(@n_cols)

            while !pos_OK(r_row, r_col) || !empty_pos(r_row, r_col)
                r_row = Dice.random_pos(@n_rows)
                r_col = Dice.random_pos(@n_cols)
            end

            [r_row, r_col]
        end

        # Puts a player in the labyrinth at the specified position.
        # @param old_row [Integer] the row index of the player's old position
        # @param old_col [Integer] the column index of the player's old position
        # @param row [Integer] the row index of the player's new position
        # @param col [Integer] the column index of the player's new position
        # @param player [Player] the player to put
        # @return [Monster] the monster in the new position, or nil if there is no monster
        def put_player_2D (old_row, old_col, row, col, player)
            output = nil
            if self.can_step_on(row, col)
                if self.pos_OK(old_row, old_col)
                    p = @players[old_row][old_col]
                    if p == player
                        self.update_old_pos(old_row, old_col)
                        @players[old_row][old_col]=nil
                    end
                end

                if self.monster_pos(row, col)
                    @labyrinth[row][col] = @@COMBAT_CHAR
                    output = @monsters[row][col]
                else
                    @labyrinth[row][col] = player.number
                end

                @players[row][col]=player
                player.set_pos(row,col)

            end

            output
        end

        public
        # Updates the player in the labyrinth.
        # @param player [Player] the player to update
        # @return [Player] the player in the new position, or nil if there is no player
        def update_player(player)
            if @players[player.row][player.col] != nil
                @players[player.row][player.col] = player
            end
        end
    end
end

